### PR TITLE
Add WorkCluster class for merging works (dedupe, suggestions, or merge from admin)

### PR DESCRIPTION
**Pull request by jilljenn**
Ongoing work.

### CODE REVIEW

**Comment by jilljenn** [comment]
TROP CHIANT.

- J'ai l'impression que si on delete un WorkCluster ses Work sont supprimés aussi
- Ajouté un champ `redirect` à Work qui n'est pas affiché dans l'admin, problème : lorsqu'on merge les works 1 et 2 et que 2 redirige vers 1, au lieu d'afficher « La fusion 1-2 s'est déroulée avec succès » ça n'affiche plus que « La fusion 1 s'est déroulée avec succès », bah oui 2 a été filtré dans le queryset ! —__—
- Peut-être que le reviewer n'appréciera pas ma gestion des réponses dans les actions `merge` et `trigger_merge` ?

**Comment by RaitoBezarius** [comment]
→ 2, why are you filtering out works with `redirect = null`?
You should make use of the redirection in the `work` view by checking if the field is non-null and do a permanent redirection.

File: mangaki/mangaki/admin.py [hunk 1]
```diff
@@ -8,13 +8,126 @@
...
 from mangaki.utils.anidb import AniDB
 from mangaki.utils.db import get_potential_posters
 
 from collections import defaultdict
 from enum import Enum
+from datetime import datetime
+
+
+@transaction.atomic  # In case trouble happens
+def merge_works(request, works_to_merge, admin_queryset, from_cluster):
```
Comment (RaitoBezarius) [inline mangaki/mangaki/admin.py]: Could we break down the logic of this function into smaller chunks?

Comment (jilljenn) [inline mangaki/mangaki/admin.py]: Yes but if so, should I add a file `merge.py` to `utils`?

**Comment by jilljenn** [comment]
I'm filtering `redirect != Null` entries in the **admin** because:

- I should not redirect an admin entry towards another admin entry;
- it is boring to see duplicates in the admin list you're browsing while you just merged them.

File: mangaki/mangaki/admin.py (hunk 1 above)
Comment (Elarnon) [inline mangaki/mangaki/admin.py]: I think it's fine to just have helper functions here in `admin.py`.

File: mangaki/mangaki/admin.py [hunk 2]
```diff
@@ -8,13 +8,126 @@
...
+            work_to_redirect.save()
+        kept_work.save()
+        cluster.resulting_work = kept_work
+        cluster.merged_on = datetime.now()
+        cluster.status = 'accepted'
+        cluster.save()
+        return None
+
+    do_not_compare = ['sum_ratings', 'nb_ratings', 'nb_likes', 'nb_dislikes', 'controversy']
+    priority = {
```
Comment (Elarnon) [inline mangaki/mangaki/admin.py]: It would probably make sense to store those directly on the MergeType enum.

```python
class MergeType(Enum):
    INFO_ONLY = (0, 'black')
    JUST_CONFIRM = (1, 'green')
    CHOICE_REQUIRED = (2, 'red')

    def __init__(self, priority, row_color):
        self.priority = priority
        self.row_color = row_color
```

File: mangaki/mangaki/admin.py [hunk 3]
```diff
@@ -76,6 +191,10 @@ class WorkAdmin(admin.ModelAdmin):
         'controversy',
     )
 
+    def get_queryset(self, request):
+        qs = super().get_queryset(request)
+        return qs.filter(redirect__isnull=True)  # Exclude phantom works
```
Comment (Elarnon) [inline mangaki/mangaki/admin.py]: Doesn't it make sense for the default `Work` queryset to exclude those so-called phantom works, with explicit overrides (via a `Work.all_objects` or some such) when we really want to see them appear -- which would probably be only when viewing a `Work` and redirecting to its canonical version?

File: mangaki/mangaki/admin.py (hunk 1 above)
Comment (Elarnon) [inline mangaki/mangaki/admin.py]: Could we have some documentation on the parameters though (in particular, what is `adin_queryset`?)

File: mangaki/mangaki/admin.py [hunk 4]
```diff
@@ -149,103 +268,11 @@ def make_sfw(self, request, queryset):
...
-            'template_rows': template_rows,
-            'rating_samples': rating_samples,
-            'queryset': queryset,
-            'opts': self.model._meta,
-            'action_checkbox_name': helpers.ACTION_CHECKBOX_NAME
-        }
-        return TemplateResponse(request, 'admin/merge_selected_confirmation.html', context)
+        response = merge_works(request, queryset, queryset, from_cluster=False)
+        if response is None:
+            self.message_user(request, "La fusion %s a bien été effectuée." % '-'.join(str(work.id) for work in queryset))
```
Comment (Elarnon) [inline mangaki/mangaki/admin.py]: I'm not sure it makes that much sense to display the IDs here. Here is how I would (mostly) write this function:

```python
from django.utils.html import format_html  # At top of file :-)

num_to_merge = len(queryset)  # Could possibly be returned by merge_works
response = merge_works(request, queryset, queryset, from_cluster=False)
if response is None:
    final_work = queryset.get()  # Should definitely be returned by merge_works 
    self.message_user(request, format_html(
        "La fusion de {:d} œuvres vers <a href="{:s}">{:s}</a> a été effectuée." % (
            num_to_merge, final_work.get_absolute_url(), final_work.title)))
return response
```

File: mangaki/mangaki/admin.py [hunk 5]
```diff
@@ -304,6 +331,39 @@ class TaggedWorkAdmin(admin.ModelAdmin):
...
+            message_bit = "1 cluster"
+        else:
+            message_bit = "%s clusters" % rows_updated
+        self.message_user(request, "Le rejet de %s a été réalisé avec succès." % message_bit)
+
+    reject.short_description = "Rejeter les clusters sélectionnés"
+
+    def get_work_titles(self, obj):
+        if obj.works.exists():
+            return '<ul>' + ''.join('<li>%s (<a href="/admin/mangaki/work/%d/change/">%d</a>)</li>' % (work.title, work.id, work.id) for work in obj.works.all()) + '</ul>'
```
Comment (Elarnon) [inline mangaki/mangaki/admin.py]: Please use https://docs.djangoproject.com/en/dev/ref/utils/#django.utils.html.format_html_join and https://docs.djangoproject.com/en/dev/ref/contrib/admin/#reversing-admin-urls here.

```python
if obj.works.exists():
    def get_admin_url(work):
        return reverse('admin:mangaki_work_change', args=(work.id,))
    return (
        '<ul>' +
        format_html_join('', '<li>{} (<a href="{}">{}</a>)</li>',
            ((work.title, get_admin_url(work), work.id) for work in obj.works.all())) +
        '</ul>'
    )
else:
    return '(all deleted)'
```

File: mangaki/mangaki/admin.py (hunk 2 above)
Comment (jilljenn) [inline mangaki/mangaki/admin.py]: Done.

File: mangaki/mangaki/admin.py (hunk 3 above)
Comment (jilljenn) [inline mangaki/mangaki/admin.py]: Yes but I can't make it. Can you? I mean, how to add extra elements when the whole source is filtered?

Comment (Elarnon) [inline mangaki/mangaki/admin.py]: Just define two managers `Work`, see https://docs.djangoproject.com/en/1.11/topics/db/managers/#modifying-a-manager-s-initial-queryset

NB: Since we use a custom QuerySet for Works, you should use `from_queryset`; something like the following:

```python
class FilteredWorkManager(models.Manager):
    def get_queryset(self):
        return super().get_queryset().filter(redirect__isnull=True)

class Work(models.Model):
    ...
    objects = FilteredWorkManager.from_queryset(WorkQuerySet)()
    all_objects = WorkQuerySet.as_manager()
```

Comment (jilljenn) [inline mangaki/mangaki/admin.py]: Done.

File: mangaki/mangaki/admin.py (hunk 4 above)
Comment (jilljenn) [inline mangaki/mangaki/admin.py]: Done.

File: mangaki/mangaki/admin.py (hunk 3 above)
Comment (jilljenn) [inline mangaki/mangaki/admin.py]: J'ai eu des soucis avec les ManyToMany.

    If you override the get_queryset() method and filter out any rows, Django will return incorrect results. Don’t do that. A manager that filters results in get_queryset() is not appropriate for use as a base manager.

On fait toujours ça ?

Comment (jilljenn) [inline mangaki/mangaki/admin.py]: In [3]: WorkCluster.objects.get(id=10).works.all()
    Out[3]: <QuerySet [<Work: Patema Inverted>]>

    In [4]: WorkCluster.objects.filter(id=10).values_list('works')
    Out[4]: <QuerySet [(15427,), (15428,)]>

:'( C'est toujours filtré.

Comment (jilljenn) [inline mangaki/mangaki/admin.py]: Ah ben c'était `default_manager_name` qu'il fallait. J'ai mis `all_objects` avant `objects` et c'est mieux.

File: mangaki/mangaki/admin.py (hunk 5 above)
Comment (jilljenn) [inline mangaki/mangaki/admin.py]: Done.

**Comment by jilljenn** [comment]
Fixes #98.

File: mangaki/mangaki/admin.py (hunk 3 above)
Comment (Elarnon) [inline mangaki/mangaki/admin.py]: Ah oui zut j'avais oublié ce truc désolé :-/ That's the right fix.

Comment (Elarnon) [inline mangaki/mangaki/admin.py]: I think this `get_queryset` function in the admin can be removed now?

Comment (Elarnon) [inline mangaki/mangaki/admin.py]: Actually the behavior is inconsistent with the doc since Django seems to be using the default manager for related object accesses instead of the base manager... I'm investigating.

Comment (Elarnon) [inline mangaki/mangaki/admin.py]: Looks like this is indeed only used for one-to-one fields due to historical reasons, which is unfortunate. I think I'd prefer using `WorkCluster.objects.get(id=10).works(manager='all_objects').all()` instead, since `WorkCluster` is "special" in the sense that it essentially the only class that should see the removed `Works`, we want them to appear as non-existent for other objects. Not 100% sure though, I think the manual solution works for now -- I'd appreciate a comment there explaining that we use manual filtering because ManyToManyField doesn't honor the `default_manager`.

File: mangaki/mangaki/admin.py [hunk 6]
```diff
@@ -1,20 +1,136 @@
...
+
+def redirect_related_objects(works_to_merge, final_work):
+    for work_to_redirect in works_to_merge:
+        if work_to_redirect.id == final_work.id:
+            continue
+        for rating_to_redirect in work_to_redirect.rating_set.all():
+            try:
+                kept_rating = Rating.objects.get(work_id=final_work.id, user_id=rating_to_redirect.user_id)
+                if kept_rating.date and rating_to_redirect.date and kept_rating.date < rating_to_redirect.date:  # Kept rating is not the latest given
+                    kept_rating.choice = rating_to_redirect.choice  # Update the kept rating
```
Comment (RaitoBezarius) [inline mangaki/mangaki/admin.py]: Is `kept_rating` saved?

File: mangaki/mangaki/admin.py [hunk 7]
```diff
@@ -1,20 +1,136 @@
...
+    final_work.save()
+
+
+def redirect_related_objects(works_to_merge, final_work):
+    for work_to_redirect in works_to_merge:
+        if work_to_redirect.id == final_work.id:
+            continue
+        for rating_to_redirect in work_to_redirect.rating_set.all():
+            try:
+                kept_rating = Rating.objects.get(work_id=final_work.id, user_id=rating_to_redirect.user_id)
```
Comment (RaitoBezarius) [inline mangaki/mangaki/admin.py]: We should optimise the queries, maybe `kept_ratings = Rating.objects.filter(work_id=final_work.id, user_id__in=all_user_ids)` and iterate over it.

File: mangaki/mangaki/admin.py [hunk 8]
```diff
@@ -1,20 +1,136 @@
...
+                rating_to_redirect.save()
+        for staff_to_redirect in work_to_redirect.staff_set.all():
+            if Staff.objects.filter(work_id=final_work.id, artist_id=staff_to_redirect.artist_id, role_id=staff_to_redirect.role_id).exists():  # This staff is already in the kept work
+                staff_to_redirect.delete()
+            else:
+                staff_to_redirect.work_id = final_work.id  # Safely transfer the staff to the kept work
+                staff_to_redirect.save()
+        final_work.genre.add(*work_to_redirect.genre.all())
+        Trope.objects.filter(origin_id=work_to_redirect.id).update(origin_id=final_work.id)
+        for model in [WorkTitle, TaggedWork, Suggestion, Recommendation, Pairing, Reference, ColdStartRating]:
```
Comment (RaitoBezarius) [inline mangaki/mangaki/admin.py]: Nit: tuple looks better here.

File: mangaki/mangaki/tests/test_merge.py [hunk 9]
```diff
@@ -0,0 +1,40 @@
+from django.test import TestCase
+from django.core.urlresolvers import reverse
+from django.contrib.auth import get_user_model
+from django.contrib import admin
+from mangaki.models import Work, Editor, Category, Studio, WorkCluster
+
+
+class RecoTest(TestCase):
```
Comment (RaitoBezarius) [inline mangaki/mangaki/tests/test_merge.py]: `MergeTest` you meant?

File: mangaki/mangaki/admin.py (hunk 7 above)
Comment (jilljenn) [inline mangaki/mangaki/admin.py]: You're right! I should, sorry.

File: mangaki/mangaki/admin.py (hunk 6 above)
Comment (jilljenn) [inline mangaki/mangaki/admin.py]: No :'(

File: mangaki/mangaki/tests/test_merge.py (hunk 9 above)
Comment (jilljenn) [inline mangaki/mangaki/tests/test_merge.py]: Oui.

File: mangaki/mangaki/admin.py (hunk 8 above)
Comment (jilljenn) [inline mangaki/mangaki/admin.py]: Sorry but I might do a set again O:-)

**Comment by jilljenn** [comment]
Ça tarde mais je vais faire un test qui vérifie que le nombre de requêtes faites est en dessous d'un certain seuil xD

File: mangaki/mangaki/admin.py (hunk 3 above)
Comment (jilljenn) [inline mangaki/mangaki/admin.py]: Done: `# Equivalent to default_manager_name = 'all_objects', because first in the list`.

File: mangaki/mangaki/tests/test_merge.py [hunk 10]
```diff
@@ -0,0 +1,62 @@
...
+        self.users = []
+        for username in 'ABCD':
+            self.users.append(get_user_model().objects.create_user(username=username, password='test'))
+
+        today = date.today()
+        yesterday = date.today() - timedelta(1)
+        tomorrow = date.today() + timedelta(1)
+
+        anime = Category.objects.get(slug='anime')
+        Work.objects.bulk_create([Work(title='Sangatsu no Lion', category=anime) for _ in range(10)])
```
Comment (RaitoBezarius) [inline mangaki/mangaki/tests/test_merge.py]: Unlimited Duplicates Works. Shouldn't we create *different* animes and assume that duplicates may throw an `IntegrityError` ?

Comment (jilljenn) [inline mangaki/mangaki/tests/test_merge.py]: These are different. They just contain the same title because I was bored. It's okay, different anime can unfortunately have the same title :)

**Comment by jilljenn** [comment]
## Let me explain what is happening here.

- Ryan noticed that I was doing too many queries.
- I made a test where an admin wants to merge 10 works with 1 rating.
- The number of queries was 142.
- The number queries is now `self.assertNumQueries(36)`.
- Even with extra ratings with the following edge cases:
    - Rating of the canonical work should be replaced with more recent ratings on works to be merged.
    - Staff pairs (artist, role) should be deleted because already present in the canonical work.

## Note that:
 
- `bulk_create` returns objects only with PostgreSQL. If we change to SQLite for faster tests, errors will happen.
- I am not sure what models should/are deleted when a model is deleted (link with `CASCADE`)?
- Maybe ratings should have a `DateTime`, not a `Date`.
- Ideally, related objects should be fetched using `Work._meta.get_fields()` (in recent versions of Django). But in these objects, the `Work` field does not always have the same attribute (it can be `origin` in `Trope`, `album` in `Track`), so I don't know how to do this elegantly. Should we prefetch every related object in order to prepare the renaming? Maybe this question is general enough to be asked on Stack Overflow.

## This PR was really boring because:

- We overrided the default manager because we don't want duplicates to appear in the admin. So it requires understanding the difference between `_default_manager` and `_base_manager`. (I can't recall.) Anyway, please do not switch the lines `all_objects` and `objects` in `models.py`.
- It is hard to fake the date of a rating for testing. `save()` triggers `auto_now`, `update_or_create` does the same, so the only way to do this is `update`. So we should create the objects, then update them. It cost me a lot of time to understand this.

File: mangaki/mangaki/admin.py [hunk 11]
```diff
@@ -1,20 +1,149 @@
...
+        kept_rating_ids.append(get_id_of_rating[(user_id, date)])
+    Rating.objects.filter(work__in=works_to_merge).exclude(id__in=kept_rating_ids).delete()
+    Rating.objects.filter(id__in=kept_rating_ids).update(work_id=final_work.id)
+
+
+def redirect_staff(works_to_merge, final_work):
+    final_work_staff = set()
+    kept_staff_ids = []
+    # Only one query: put final_work's Staff objects first in the list
+    for staff_id, work_id, artist_id, role_id in Staff.objects.filter(work__in=works_to_merge).annotate(belongs_to_final_work=Case(When(work_id=final_work.id, then=Value(1)), default=Value(0), output_field=IntegerField())).order_by('-belongs_to_final_work').values_list('id', 'work_id', 'artist_id', 'role_id'):
```
Comment (RaitoBezarius) [inline mangaki/mangaki/admin.py]: We really have to cut this line, it's too long.
It's okay to create variables.

Comment (jilljenn) [inline mangaki/mangaki/admin.py]: Done!

**Comment by Elarnon** [comment]
> Ideally, related objects should be fetched using Work._meta.get_fields() (in recent versions of Django). But in these objects, the Work field does not always have the same attribute (it can be origin in Trope, album in Track), so I don't know how to do this elegantly. Should we prefetch every related object in order to prepare the renaming? Maybe this question is general enough to be asked on Stack Overflow.

You can query based on the type of the field, the name is not relevant. I have some prototype work for this from September lying around somewhere, I can dig it up if needed.

> We overrided the default manager because we don't want duplicates to appear in the admin. So it requires understanding the difference between _default_manager and _base_manager. (I can't recall.) Anyway, please do not switch the lines all_objects and objects in models.py.

I'm not sure why you didn't want the explicit version of assigning `_default_manager` by hand instead of relying on the impliciteness of "the default manager is the first defined" then.

> It is hard to fake the date of a rating for testing. save() triggers auto_now, update_or_create does the same, so the only way to do this is update. So we should create the objects, then update them. It cost me a lot of time to understand this.

I believe fixtures are the right solution for this problem.

> I am not sure what models should/are deleted when a model is deleted (link with CASCADE)?

Yes, this is something we should think about, and yes [it has a link with CASCADE](https://docs.djangoproject.com/en/1.11/ref/models/fields/#django.db.models.ForeignKey.on_delete). Some of the `ForeignKey`s should be `on_delete=PROTECT` at the very least -- for instance, `Ratings`, because all the ratings for a `Work` should have been either re-pointed to the de-duplicated `Work` or carefully deleted before deleting a `Work`.
